src/common/ad_tx.c

```diff
@@ -0,0 +1,480 @@
+/**
+ * (C) Copyright 2022 Intel Corporation.
+ *
+ * SPDX-License-Identifier: BSD-2-Clause-Patent
+ */
+#define D_LOGFAC	DD_FAC(common)
+
+#include <daos_srv/ad_mem.h>
+
+enum {
+	ACT_UNDO = 0,
+	ACT_REDO = 1,
+};
+
+#define AD_TX_ACT_ADD(tx, it, undo_or_redo)						\
+	do {										\
+		if (undo_or_redo == ACT_UNDO)						\
+			d_list_add(&it->it_link, &tx->tx_undo);				\
+		else									\
+			d_list_add_tail(&it->it_link, &tx->tx_redo);			\
+	} while (0)
+
+#define AD_TX_ACT_DEL(it)	d_list_del(it)
+
+/** allocate umem_act_item, if success the it_link and it_act.ac_opc will be init-ed */
+#define D_ALLOC_ACT(it, opc, size)							\
+	do {										\
+		if (opc == UMEM_ACT_COPY)						\
+			D_ALLOC(it, offsetof(struct umem_act_item,			\
+					     it_act.ac_copy.payload[size]));		\
+		else if (opc == UMEM_ACT_CSUM)						\
+			D_ALLOC(it, offsetof(struct umem_act_item,			\
+					     it_act.ac_csum.csum[size]));		\
+		else									\
+			D_ALLOC_PTR(it);						\
+		if (likely(it != NULL)) {						\
+			D_INIT_LIST_HEAD(&it->it_link);					\
+			it->it_act.ac_opc = opc;					\
+		}									\
+	} while (0)
```

Comment:
I'm bit worried about the memory allocations, now basically every memory operation (over tmpfs) will cause a memory allocation, shouldn't we just allocate a vector to put all 'actions' in (the vector could grow on demand)? That could save lots of small allocations.

Reply:
the action allocation is using standard dram allocator, it seems commonly already with thread-cache slab in the allocator.
with our own slab vector for actions can be considered to add in following patch, although I am not sure if can get big benefit.